### ISSUE TITLE
FX11 support for shaders

### ISSUE THREAD

**Issue by Swartz27** [xrOxygen/xray-oxygen#97]
I'm working on a variety of things that I plan on doing a pull request for (some good and balanced visuals for all three renderers), but anyway, as someone that knows very little C++, it sucks as a shader author not really understanding how to implement everything needed, most importantly: techniques/multi-passes in one shader.

One possible solution could be FX11 (and FX9&10 too I suppose). 

Basically: if you were to add something like that, the sky would be the limit in terms of very quickly adding of high quality effects with shader authors having a much greater control over what they can do inside a shader.

Example 1: I know exactly how to add SMAA via a shader. However, due to the fact that I need to make multiple passes inside the same shader instead of being limit to one pass that has to be named "main", I cannot do this without touching the source code, which isn't so much problematic due to my lack of C++ understanding; it's more that GSC just had a very bizarre way of setting things up.
So: a feature we all would like that would take a matter of minutes to add with any other engine that has technique/multipass support, whereas with Stalker it can be a huge pain in the ass that can take a long time.

Example 2: Due to being able to customize sampler states, being able to create render targets inside shader (including to existing texture files), I could very quickly give you guys Voxel AO, or hell, why not GI too? I have great shader code for doing this, I just need a technique/multi-pass ability to add it (and more importantly, avoid having to touch the code when adding brand new shaders and instead have them auto-detected or something [AND MOST IMPORTANTLY: Not just PS, and VS, but CS, and GS are crucial not just for great graphics, but in the case of CS you can greatly optimize and squeeze more performance out of items you would normally put in a pixel shader].

tl;dr: FX11 would be nice, or something similar.

**Comment by ForserX** [comment]
I have the opposite. There is no understanding of the shader system. Because of what I can not really help you.

I'll try at leisure to understand what FX11 is, but I do not promise anything.

In any case, you can send PR, whether it's a shader fix, a new effect, or something else. We will support and try to drive it into the C ++ sampler.

**Comment by Swartz27** [comment]
Luckily FX11 is actually C++: you wouldn't have to touch shader code, just the renderer code in the source.

As for trouble understanding the shader system, I'm pretty well versed in this stuff at this point, so even if I currently have no means of adding certain things C++ side, if you guys can do that part I can take requests and at least prepare some new shaders for usage. I can include them in a PR along with the other modifications I'm doing: I can add them to a specially marked sub-folder or something that includes commenting inside the shader on exactly what I need done C++ side to get those shaders working.

**Comment by ForserX** [comment]
Okay.

**Comment by ForserX** [comment]
Src added.
